### PR TITLE
allow users to pass in credentials to the AwsS3Connector in the config

### PR DESCRIPTION
**Pull request by veilig2000**
creates ability to utilize credential providers for the S3 adapter.

http://docs.aws.amazon.com/aws-sdk-php/v3/guide/guide/credentials.html

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
No change is needed.

**Comment by GrahamCampbell** [comment]
Oh, right, I see.

**Comment by GrahamCampbell** [comment]
Once this has tests, I'll merge it.

**Comment by GrahamCampbell** [comment]
Does credentials only support key and secret? If so, then it's uneeded.

**Comment by GrahamCampbell** [comment]
I have no intentional to purposefully match up with the internal config for aws. Any matchings are coincidental.

**Comment by veilig2000** [comment]
no,  you can pass in providers dealing with environment settings, ini files, instance profiles for EC2 (i.e. IAM's). but it also takes a key and secret as a basic input as well - which was the easiest one to test. Specifically b/c I didn't want to deal w/ mocking a promise that guzzle would return, and/or mock the aws client so I could access the credentials

http://docs.aws.amazon.com/aws-sdk-php/v3/guide/guide/credentials.html

**Comment by veilig2000** [comment]
basically - my my use case is to grant my EC2 boxes a role which allows access to a S3 bucket, and then memoize the credentials in the code with a promise and the AWS client can retrieve the permissions from the instance when needed. That way I don't have to worry about keys / secrets in source code or deployments.

**Comment by GrahamCampbell** [comment]
Could you provide a testcase for that usage please. That's what I was originally getting at. :)

**Comment by veilig2000** [comment]
ah, ya sorry.  I'm gonna have to sit down and look at that. I need to look at their docs and see how to mock their providers so I can verify I get what I expect in the test then. Like I said, passing in the hard coded credentials like I am currently doing in my test was the easiest route, but when you track down the credentials from the aws client, they're returned in a guzzle http promise as a private value and without reflection in the test (yucky) there was no way to verify them.

That being said, I'm not completely sold writing a test case to verify whats passed into the credentials index of the config actually works is the responsibility of your code.  The tests already show that including a credentials index still works correctly.

File: tests/Adapters/AwsS3ConnectorTest.php [hunk 1]
```diff
@@ -138,6 +138,33 @@ public function testConnectWithEverything()
...
+     * Using the provider to grab credentials from environment variables
+     */
+    public function testConnectWithCredentials()
+    {
+        $connector = $this->getAwsS3Connector();
+
+        putenv('AWS_ACCESS_KEY_ID=your-access-key');
+        putenv('AWS_SECRET_ACCESS_KEY=your-secret');
+
+        $provider = \Aws\Credentials\CredentialProvider::defaultProvider();
```
Comment (GrahamCampbell) [inline tests/Adapters/AwsS3ConnectorTest.php]: Please import this.

File: tests/Adapters/AwsS3ConnectorTest.php [hunk 2]
```diff
@@ -138,6 +138,33 @@ public function testConnectWithEverything()
...
     /**
+     * Testing credential providers.
+     *
+     * Using the provider to grab credentials from environment variables
+     */
+    public function testConnectWithCredentials()
+    {
+        $connector = $this->getAwsS3Connector();
+
+        putenv('AWS_ACCESS_KEY_ID=your-access-key');
```
Comment (GrahamCampbell) [inline tests/Adapters/AwsS3ConnectorTest.php]: Please avoid calling this function.

**Comment by GrahamCampbell** [comment]
I think I see where this is going, and I don't intend to support this. You can get that behaviour to work with some clever code in the config file.

**Comment by veilig2000** [comment]
I'm not exactly why you wouldn't support functionality provided by AWS.  I also don't see how you can get this behavior to work w/ code in the config since it literally needs to be passed to the `credentials` index of the config array and you clearly overwrite that index https://github.com/GrahamCampbell/Laravel-Flysystem/blob/master/src/Adapters/AwsS3Connector.php#L69

About the only way I see you would be able to make this work would be to substitute either the key or secret value for credentials, but even then, I doubt its going to work the way you have it setup. On top of the fact that its a convoluted approach.

**Pull request closed without merging**